samza-core/src/test/java/org/apache/samza/zk/TestZkBarrierForVersionUpgrade.java

```diff
@@ -18,205 +18,221 @@
       e.printStackTrace();
     }
-    Assert.assertTrue("Barrier failed to complete within test timeout.", result);
+    assertTrue("Barrier failed to complete within test timeout.", result);
+
+    List<String> children = zkUtils.getZkClient().getChildren(barrierId + "/barrier_1/barrier_participants");
+    barrierState = zkUtils.getZkClient().readData(barrierId + "/barrier_1/barrier_state");
+    assertEquals(State.DONE, barrierState);
+    assertNotNull(children);
+    assertEquals("Unexpected barrier state. Didn't find two processors.", 2, children.size());
+    assertEquals("Unexpected barrier state. Didn't find the expected members.", processors, children);
+    assertEquals(2, stateChangedCalled.get());
+  }
+
+  @Test
+  public void testZkBarrierForVersionUpgradeWithTimeOut() {
+    String barrierId = String.format("%s/%s", zkUtils1.getKeyBuilder().getRootPath(), RandomStringUtils.randomAlphabetic(4));
+    List<String> processors = ImmutableList.of("p1", "p2", "p3");
+
+    AtomicInteger timeoutStateChangeCalled = new AtomicInteger(0);
+    CountDownLatch latch = new CountDownLatch(2);
+    TestZkBarrierListener listener = new TestZkBarrierListener(latch, timeoutStateChangeCalled, State.TIMED_OUT);
+
+    ZkBarrierForVersionUpgrade processor1Barrier = new ZkBarrierForVersionUpgrade(barrierId, zkUtils, listener);
+    ZkBarrierForVersionUpgrade processor2Barrier = new ZkBarrierForVersionUpgrade(barrierId, zkUtils1, listener);
 
+    processor1Barrier.create(BARRIER_VERSION, processors);
+
+    processor1Barrier.join(BARRIER_VERSION, "p1");
+    processor2Barrier.join(BARRIER_VERSION, "p2");
+
+    processor1Barrier.expire(BARRIER_VERSION);
+    boolean result = false;
     try {
-      List<String> children = zkUtils.getZkClient().getChildren(barrierId + "/barrier_v1/barrier_participants");
-      Assert.assertNotNull(children);
-      Assert.assertEquals("Unexpected barrier state. Didn't find two processors.", 2, children.size());
-      Assert.assertEquals("Unexpected barrier state. Didn't find the expected members.", processors, children);
-    } catch (Exception e) {
-      // no-op
+      result = latch.await(10000, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      e.printStackTrace();
     }
-    Assert.assertEquals(2, stateChangedCalled.get());
+    assertTrue("Barrier Timeout test failed to complete within test timeout.", result);
+
+    List<String> children = zkUtils.getZkClient().getChildren(barrierId + "/barrier_1/barrier_participants");
+    State barrierState = zkUtils.getZkClient().readData(barrierId + "/barrier_1/barrier_state");
+    assertEquals(State.TIMED_OUT, barrierState);
+    assertNotNull(children);
+    assertEquals("Unexpected barrier state. Didn't find two processors.", 2, children.size());
+    assertEquals("Unexpected barrier state. Didn't find the expected members.", ImmutableList.of("p1", "p2"), children);
+    assertEquals(2, timeoutStateChangeCalled.get());
```

Comment:
Could this be flaky in this scenario?

- thread-1
await on the latch
check counter==2

- zk-client thread
listener.onEvent() is called for processor-1
processor1 increments counter
processor1 counts down the latch
listener.onEvent() is called for processor-2
processor2 increments the counter

Could the counter-check in thread-1 succeed/fail depending on whether the ZkClient thread got to deliver the event for processor2?

Would be simpler to use a single latch with count = 2 or use separate listener instances